File: sections/17_topological_sorting_dag_dp/problems/c_project_schedule/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<long long> duration(n);
    for (long long &x : duration) cin >> x;
    vector<vector<int>> adj(n);
    vector<int> indeg(n, 0);
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        --a;
        --b;
        adj[a].push_back(b);
        indeg[b]++;
    }

    vector<long long> earliest_start(n, 0), finish = duration;
    queue<int> q;
    for (int i = 0; i < n; i++) {
        if (indeg[i] == 0) q.push(i);
    }

    int processed = 0;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        processed++;
        finish[u] = earliest_start[u] + duration[u];
        for (int v : adj[u]) {
            earliest_start[v] = max(earliest_start[v], finish[u]);
            indeg[v]--;
            if (indeg[v] == 0) q.push(v);
        }
    }

    if (processed != n) {
        cout << "IMPOSSIBLE\n";
    } else {
        for (int i = 0; i < n; i++) {
            if (i) cout << ' ';
            cout << finish[i];
        }
        cout << '\n';
    }
    return 0;
}

File: sections/17_topological_sorting_dag_dp/problems/c_project_schedule/solution.py
from collections import deque
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    m = next(it)
    duration = [next(it) for _ in range(n)]
    adj = [[] for _ in range(n)]
    indeg = [0] * n
    for _ in range(m):
        a = next(it) - 1
        b = next(it) - 1
        adj[a].append(b)
        indeg[b] += 1

    earliest_start = [0] * n
    finish = duration[:]
    q = deque(i for i in range(n) if indeg[i] == 0)
    processed = 0
    while q:
        u = q.popleft()
        processed += 1
        finish[u] = earliest_start[u] + duration[u]
        for v in adj[u]:
            if earliest_start[v] < finish[u]:
                earliest_start[v] = finish[u]
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)

    if processed != n:
        print("IMPOSSIBLE")
    else:
        print(" ".join(map(str, finish)))


if __name__ == "__main__":
    main()

